tiatoolbox/models/architecture/sccnn.py

```diff
@@ -0,0 +1,372 @@
+                ),
+                nn.ReLU(),
+            )
+
+            return nn.ModuleDict(module_dict)
+
+        def spatially_constrained_layer1(
+            in_channels, out_channels
+        ) -> torch.nn.ModuleDict:
+            """Spatially constrained layer.
+
+            Takes fully connected layer and returns outputs for creating probability
+            map for the output.
+
+            Args:
+                in_channels (int):
+                    Number of channels in input.
+                out_channels (int):
+                    Number of required channels in output.
+
+            Returns:
+                torch.nn.ModuleDict:
+                    Module dictionary.
+
+            """
+            module_dict = OrderedDict()
+            module_dict["conv1"] = nn.Sequential(
+                nn.Conv2d(
+                    in_channels,
+                    out_channels,
+                    kernel_size=(1, 1),
+                    stride=(1, 1),
+                    padding=0,
+                    bias=True,
+                ),
+                nn.Sigmoid(),
+            )
+            return nn.ModuleDict(module_dict)
+
+        module_dict = OrderedDict()
+        module_dict["l1"] = conv_act_branch(num_input_channels, 30, 2)
+        module_dict["pool1"] = nn.MaxPool2d(2, padding=0)
+        module_dict["l2"] = conv_act_branch(30, 60, 2)
+        module_dict["pool2"] = nn.MaxPool2d(2, padding=0)
+        module_dict["l3"] = conv_act_branch(60, 90, 3)
+        module_dict["l4"] = conv_act_branch(90, 1024, 5)
+        module_dict["dropout1"] = nn.Dropout2d(p=0.5)
+        module_dict["l5"] = conv_act_branch(1024, 512, 1)
+        module_dict["dropout2"] = nn.Dropout2d(p=0.5)
+        module_dict["sc"] = spatially_constrained_layer1(512, 3)
+
+        self.layer = nn.ModuleDict(module_dict)
+
+    def forward(self, input_tensor: torch.Tensor):  # skipcq: PYL-W0221
+        """Logic for using layers defined in init.
+
+        This method defines how layers are used in forward operation.
+
+        Args:
+            input_tensor (torch.Tensor):
+                Input images, the tensor is in the shape of NCHW.
+
+        Returns:
+            torch.Tensor:
+                Output map for cell detection. Peak detection should be applied
+                to this output for cell detection.
+
+        """
+
+        def conv_act_branch(
+            layer: torch.nn.Module, in_tensor: torch.Tensor
+        ) -> torch.Tensor:
+            """Applies Convolution and Activation to the input tensor.
+
+            Args:
+                layer (torch.nn.ModuleDict):
+                    Torch layer as ModuleDict.
+                in_tensor (torch.Tensor):
+                    Input Tensor.
+            Returns:
+                torch.Tensor:
+                    Torch Tensor after applying convolution and activation.
+
+            """
+            return layer["conv1"](in_tensor)
+
+        def spatially_constrained_layer1(
```

Comment:
auxiliary layer are usually defined in the `__init__` function. Consider moving this and `spatially_constrained_layer2` layers function definitions to `__init__`.

Reply:
I would suggest turning this into a proper Class instead of keeping this as a function

```
class SpatiallyConstrainedLayer(nn.module):
    def forward(self,...):
        """copy content"""
```

Reply:
Moved it outside in the class definition in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/434/commits/4d6adcd9ef7f56850c4480e443b8ad15790ff8c7

Ignored class definition for `SpatiallyConstrainedLayer`

Reply:
I like @vqdang's suggestion, but happy to go with either approach.